src/community/security/oauth2-openid-connect/oauth2-openid-connect-core/src/main/java/org/geoserver/security/oauth2/OpenIdConnectAuthenticationFilter.java

```diff
@@ -61,22 +70,105 @@ public OpenIdConnectAuthenticationFilter(
+    }
+
+    @Override
+    protected String getPreAuthenticatedPrincipal(HttpServletRequest req, HttpServletResponse resp)
+            throws IOException, ServletException {
+        String result = super.getPreAuthenticatedPrincipal(req, resp);
+
+        if (result == null) {
+            return null;
+        }
+
+        OAuth2AuthenticationType type =
+                (OAuth2AuthenticationType) req.getAttribute(OAUTH2_AUTHENTICATION_TYPE_KEY);
+        if ((type != null)
+                && (type.equals(OAuth2AuthenticationType.BEARER))
+                && (bearerTokenValidator != null)) {
+            if (!((OpenIdConnectFilterConfig) filterConfig).isAllowBearerTokens()) {
+                LOGGER.log(
+                        Level.WARNING,
+                        "OIDC: received an attached Bearer token, but Bearer tokens aren't allowed!");
+                throw new IOException(
+                        "OIDC: received an attached Bearer token, but Bearer tokens aren't allowed!");
+            }
+            // we must validate
+            String accessToken =
+                    (String) req.getAttribute(OAuth2AuthenticationDetails.ACCESS_TOKEN_VALUE);
+            Map userinfoMap = (Map) req.getAttribute(OAUTH2_ACCESS_TOKEN_CHECK_KEY);
+            Jwt decodedAccessToken = JwtHelper.decode(accessToken);
+            Map accessTokenClaims = JSONObject.fromObject(decodedAccessToken.getClaims());
+            try {
+                bearerTokenValidator.verifyToken(
+                        (OpenIdConnectFilterConfig) filterConfig, accessTokenClaims, userinfoMap);
+            } catch (Exception e) {
+                throw new IOException("Attached Bearer Token is invalid", e);
+            }
+        }
+
+        return result;
     }
 
     @Override
     protected Collection<GeoServerRole> getRoles(HttpServletRequest request, String principal)
             throws IOException {
         RoleSource rs = getRoleSource();
+
+        // todo remove - security violation, but necessary for debugging
+        // If you're having issues, uncomment this to see more details of the tokens being sent
+        // These tokens are PRIVATE so ONLY do this when necessary!
+
+        //        LOGGER.log(Level.FINE, "OPEN ID CONNECT ---- ===== info for getRoles");
+        //        try {
+        //            LOGGER.log(Level.FINE, "principal: " + principal);
+        //        } catch (Exception e) {
+        //        }
+        //        try {
+        //            LOGGER.log(Level.FINE,"access token: " +
+        //                   (String)
+        // request.getAttribute(OAuth2AuthenticationDetails.ACCESS_TOKEN_VALUE));
+        //        } catch (Exception e) {
+        //        }
+        //        try {
+        //            LOGGER.log(Level.FINE, "ID token: " + (String)
+        // request.getAttribute(ID_TOKEN_VALUE));
+        //        } catch (Exception e) {
+        //        }
```

Comment:
Thanks for the comments; this is the rare case where keeping commented out code in the codebase may be worthwhile.

Reply:
One trick I do is to do an:
```
private static boolean DEBUGGING = false;

if (DEBUGGING) {
   // some stuff that would otherwise be commented out
}
```

May have to do a PMD ignore to get that passed QA checks but at least the debugging code will still be compiled.